pkg/services/scm_provider/gitlab.go

```diff
@@ -0,0 +1,137 @@
+package scm_provider
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+type GitlabProvider struct {
+	client       *gitlab.Client
+	organization string
+	allBranches  bool
+}
+
+var _ SCMProviderService = &GitlabProvider{}
+
+func newTrue() *bool {
+	b := true
+	return &b
+}
+
+func NewGitlabProvider(ctx context.Context, organization string, token string, url string, allBranches bool) (*GitlabProvider, error) {
+	// Undocumented environment variable to set a default token, to be used in testing to dodge anonymous rate limits.
+	if token == "" {
+		token = os.Getenv("GITLAB_TOKEN")
+	}
+	var client *gitlab.Client
+	if url == "" {
+		var err error
+		client, err = gitlab.NewClient(token)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		var err error
+		client, err = gitlab.NewClient(token, gitlab.WithBaseURL(url))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &GitlabProvider{client: client, organization: organization, allBranches: allBranches}, nil
+}
+
+func (g *GitlabProvider) ListRepos(ctx context.Context, cloneProtocol string) ([]*Repository, error) {
+	opt := &gitlab.ListGroupProjectsOptions{
+		ListOptions:      gitlab.ListOptions{PerPage: 100},
+		IncludeSubgroups: newTrue(),
+	}
+	repos := []*Repository{}
+	for {
+		gitlabRepos, resp, err := g.client.Groups.ListGroupProjects(g.organization, opt)
+		if err != nil {
+			return nil, fmt.Errorf("error listing projects for %s: %v", g.organization, err)
+		}
+		for _, gitlabRepo := range gitlabRepos {
+			var url string
+			switch cloneProtocol {
+			// Default to SSH if unspecified (i.e. if "").
+			case "", "ssh":
+				url = gitlabRepo.SSHURLToRepo
+			case "https":
+				url = gitlabRepo.HTTPURLToRepo
+			default:
+				return nil, fmt.Errorf("unknown clone protocol for Gitlab %v", cloneProtocol)
+			}
+
+			branches, err := g.listBranches(ctx, gitlabRepo)
+			if err != nil {
+				return nil, fmt.Errorf("error listing branches for %s/%s: %v", g.organization, gitlabRepo.Name, err)
+			}
+
+			for _, branch := range branches {
+				repos = append(repos, &Repository{
+					Organization: gitlabRepo.Namespace.FullPath,
+					Repository:   gitlabRepo.Path,
+					URL:          url,
+					Branch:       branch,
+					Labels:       gitlabRepo.TagList,
+				})
+			}
+		}
+		if resp.CurrentPage >= resp.TotalPages {
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+	return repos, nil
+}
+
+func (g *GitlabProvider) RepoHasPath(ctx context.Context, repo *Repository, path string) (bool, error) {
+	p, _, err := g.client.Projects.GetProject(repo.Organization+"/"+repo.Repository, nil)
+	if err != nil {
+		return false, err
+	}
+	_, resp, err := g.client.Repositories.ListTree(p.ID, &gitlab.ListTreeOptions{
+		Path: &path,
+		Ref:  &repo.Branch,
+	})
+	fmt.Printf("%+v\n", err)
```

Comment:
I'm guessing these `Printfs` can be removed (or converted into debug log statements, if you feel they are important enough)

Reply:
removed